src/Components/Web/src/WebRenderer.cs

```diff
@@ -51,15 +76,57 @@ protected internal int AddRootComponent([DynamicallyAccessedMembers(Component)]
 
+        /// <inheritdoc />
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _interopMethodsReference.Dispose();
+            }
+
+            base.Dispose(disposing);
+        }
+
         /// <summary>
-        /// Enables support for adding, updating, and removing root components from JavaScript.
+        /// A collection of JS invokable methods that the JS-side code can use when it needs to
+        /// make calls in the context of a particular renderer. This object is never exposed to
+        /// .NET code so is only reachable via JS.
         /// </summary>
-        /// <param name="interop">The object that provides JS-callable methods for adding, updating, and removing root components.</param>
-        /// <returns>A task representing the completion of the operation.</returns>
-        protected ValueTask InitializeJSComponentSupportAsync(JSComponentInterop interop)
+        private class WebRendererInteropMethods
         {
-            var jsRuntime = _serviceProvider.GetRequiredService<IJSRuntime>();
-            return interop.InitializeAsync(jsRuntime, this);
+            private readonly WebRenderer _renderer;
+            private readonly JsonSerializerOptions _jsonOptions;
+            private readonly JSComponentInterop _jsComponentInterop;
+
+            [DynamicDependency(DynamicallyAccessedMemberTypes.PublicMethods, typeof(WebRendererInteropMethods))]
+            public WebRendererInteropMethods(WebRenderer renderer, JsonSerializerOptions jsonOptions, JSComponentInterop jsComponentInterop)
+            {
+                _renderer = renderer;
+                _jsonOptions = jsonOptions;
+                _jsComponentInterop = jsComponentInterop;
+            }
+
+            [JSInvokable]
+            public Task DispatchEventAsync(JsonElement eventDescriptor, JsonElement eventArgs)
+            {
+                var webEventData = WebEventData.Parse(_renderer, _jsonOptions, eventDescriptor, eventArgs);
+                return _renderer.DispatchEventAsync(
+                    webEventData.EventHandlerId,
+                    webEventData.EventFieldInfo,
+                    webEventData.EventArgs);
+            }
+
+            [JSInvokable]
+            public int AddRootComponent(string identifier, string domElementSelector)
+                => _jsComponentInterop.AddRootComponent(identifier, domElementSelector);
+
+            [JSInvokable]
+            public void SetRootComponentParameters(int componentId, int parameterCount, JsonElement parametersJson)
+                => _jsComponentInterop.SetRootComponentParameters(componentId, parameterCount, parametersJson, _jsonOptions);
+
+            [JSInvokable]
+            public void RemoveRootComponent(int componentId)
+                => _jsComponentInterop.RemoveRootComponent(componentId);
```

Comment:
TODO: Change the linking config so that the JS component interop methods can get trimmed if you didn't add any JS root components.

Reply:
Done